### PR TITLE
DAC voting, base fee proposal

### PR DESCRIPTION
**Pull request by evlekht**
## Why this should be merged
This PR adds feature of DAC proposals and voting which is executed with transactions.
Also, this PR renames all DAO entries to DAC.

## How this works
There are 3 new transactions:
- AddProposalTx, it contains proposal bytes and proposer address/auth. In addition to fee burning, this tx will bond some tokens. Also, this tx must be additionally signed by proposer.
- AddVoteTx, it contains vote bytes, voter address/auth and proposalID (id of addProposalTx). This this must be additionally signed by voter.
- FinishProposalsTx is system tx. It will finish all expired (both successful and failed) and early (again both successful and failed)  proposals, tx body contains those proposal ids.

Vote and proposal fields in txs represented by bytes, not some structure or interface. This is done in order to be able to add new proposal types without needing to update codec.

Block builder will try to build block with FinishProposalsTx when there are early finished proposals or expired proposals.

Proposals execution and semantic verification logic resides in its own proposalExecutor type methods in txs/executor package. Proposals are calling this methods via visitor pattern.

Proposals and votes logic are encapsulated behind interfaces, which could be implemented by different types. This PR adds BaseFeeProposal and SimpleVote types.

SimpleVote is vote that simply points to specific proposal option index, e.g. baseFee proposal with options [ {newFee: 100}, {newFee: 200} ], and simpleVote {0}, where vote is for newFee 100 option.

BaseFeeProposal is proposal to change base fee, its options are new base fees. This proposal can only be proposed by caminoProposer address state role, which is introduced in this PR as well.

All proposals can be voted only by addresses that were active validators at the moment of addProposalTx block execution. While its not required for voters to still be active valaditors when they vote, it is required for them to still be consortium members.

### Commits ###
PR consists of 3 commits:
1) Commit with base logic except most of proposal-related tests, cause the need some actual proposal implementation.
2) Commit with simpleVote and corresponding tests
3) Commit with baseFeeProposal, corresponding tests and all tests that were excluded from (1)

## How this was tested
Unit-tests, integration tests

### CODE REVIEW

File: version/constants.go [hunk 1]
```diff
@@ -119,6 +119,13 @@ var (
 	}
 	AthensPhaseDefaultTime = time.Date(2023, time.July, 1, 8, 0, 0, 0, time.UTC)
 
+	DACPhaseTimes = map[uint32]time.Time{
```
Comment (c4t-ag) [inline version/constants.go]: I don't think that DAC is a valid name in the scheme we want to have, after A(thens) there should be B(????) and it should be something touristic. Pls. discuss with product.

File: vms/platformvm/dac/camino_change_base_fee_proposal.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	copy(updatedProposal.AllowedVoters, p.AllowedVoters[:voterAddrPos])
+	updatedProposal.AllowedVoters = append(updatedProposal.AllowedVoters[:voterAddrPos], p.AllowedVoters[voterAddrPos+1:]...)
+	// we can't use the same slice, cause we need to change its element
+	copy(updatedProposal.Options, p.Options)
+	updatedProposal.Options[vote.OptionIndex].Weight++
+	return updatedProposal, nil
+}
+
+// Will return modified proposal with added vote ignoring allowed voters, original proposal will not be modified!
+func (p *BaseFeeProposalState) ForceAddVote(voterAddress ids.ShortID, voteIntf Vote) (ProposalState, error) { //nolint:revive
```
Comment (c4t-ag) [inline vms/platformvm/dac/camino_change_base_fee_proposal.go:167]: For what is the nolint required? Maybe comment?

File: vms/platformvm/dac/camino_vote.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	return []uint32{v.OptionIndex}
+}
+
+func (*SimpleVote) Verify() error {
+	return nil
+}
+
+type SimpleVoteOption[T any] struct {
+	Value  T      `serialize:"true"`
+	Weight uint32 `serialize:"true"`
```
Comment (c4t-ag) [inline vms/platformvm/dac/camino_vote.go]: What is weight for? Maybe also short comment for struct fields.

File: vms/platformvm/locked/camino_lock.go [hunk 4]
```diff
@@ -167,6 +167,12 @@ func (out *Out) Verify() error {
 	return out.TransferableOut.Verify()
 }
 
+func (out *Out) Equal(to any) bool {
```
Comment (c4t-ag) [inline vms/platformvm/locked/camino_lock.go:171]: Where are these 2 Equal funcs are used? Just for tests?

File: vms/platformvm/txs/executor/staker_tx_verification.go [hunk 5]
```diff
@@ -289,6 +289,11 @@ func removeSubnetValidatorValidation(
 		return nil, false, err
 	}
 
+	fee, err := chainState.GetBaseFee()
```
Comment (c4t-ag) [inline vms/platformvm/txs/executor/staker_tx_verification.go:302]: Licence header (beginning of file)

File: vms/platformvm/txs/executor/camino_tx_executor.go [hunk 6]
```diff
@@ -233,8 +251,8 @@ func (e *CaminoStandardTxExecutor) AddValidatorTx(tx *txs.AddValidatorTx) error
 			tx.Outs,
 			e.Tx.Creds[:len(e.Tx.Creds)-1],
 			0,
-			e.Backend.Config.AddPrimaryNetworkValidatorFee,
-			e.Backend.Ctx.AVAXAssetID,
+			e.Backend.Config.AddPrimaryNetworkValidatorFee, // TODO@ use baseFee?
```
Comment (c4t-ag) [inline vms/platformvm/txs/executor/camino_tx_executor.go:255]: We should discuss TX per TX if we want to use base-fee. Some TX could cost lot of CAM (e.g. CreateChain), and there are TX for subnets where subnet's should be able to control fees.
Instead changing basefee directly we discussed some time a go a fee multiplicator, which could also directly control all the fees mentioned in todo here.

**Review by c4t-ag — CHANGES_REQUESTED** [review]
Nice work!
There are some minor things (comments), and some general things which should be at least discussed:
1.) If possible (not sure if it makes sense) the final PR should have at least 2 commits, the base implementation and another commit the basefee specialization. This would help to see what is required to implement a more complex one (thinking about a multi option proposal where the exit is not necessarily a simple 50% one.
2.) It should be evaluated (with @Noctunus because the idea was initially from him) to operate with a fee multiplier instead setting the basefee. This could then be applied to all spialized Fees.

File: vms/platformvm/camino_service.go [hunk 7]
```diff
@@ -1020,5 +1022,53 @@ func (s *CaminoService) GetUpgradePhases(_ *http.Request, _ *struct{}, response
...
 		response.AthensPhase = 1
 	}
+	if s.vm.Config.IsDACPhaseActivated(s.vm.state.GetTimestamp()) {
+		response.DACPhase = 1
+	}
+	return nil
+}
+
+type ConsortiumMemberValidator struct {
+	Weight                  utilsjson.Uint64 `json:"weight"`
```
Comment (c4t-ag) [inline vms/platformvm/camino_service.go]: Suggestion: ValidatorWeight

File: vms/platformvm/dac/camino_change_base_fee_proposal.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+var (
+	_ Proposal      = (*BaseFeeProposal)(nil)
+	_ ProposalState = (*BaseFeeProposalState)(nil)
+
+	errZeroFee           = errors.New("base-fee option is zero")
+	errWrongOptionsCount = errors.New("wrong options count")
+)
+
+type BaseFeeProposal struct {
+	Options []uint64 `serialize:"true"`
```
Comment (c4t-ag) [inline vms/platformvm/dac/camino_change_base_fee_proposal.go]: Pls. comment at least Options

File: vms/platformvm/dac/camino_change_base_fee_proposal.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+	return visitor.BaseFeeProposal(p)
+}
+
+type BaseFeeProposalState struct {
+	SimpleVoteOptions[uint64] `serialize:"true"`
+
+	Start              uint64        `serialize:"true"`
+	End                uint64        `serialize:"true"`
+	AllowedVoters      []ids.ShortID `serialize:"true"`
+	TotalAllowedVoters uint32        `serialize:"true"`
```
Comment (c4t-ag) [inline vms/platformvm/dac/camino_change_base_fee_proposal.go]: What is TotalAllowedVoters?

**Review by c4t-ag — CHANGES_REQUESTED** [review]
Some more comments for additional comments :-)

File: version/constants.go (hunk 1 above)
Comment (evlekht) [reply version/constants.go]: Was a placeholder. Renamed to Berlin

File: vms/platformvm/dac/camino_vote.go (hunk 3 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_vote.go]: Commented all dac pkg struct fields

File: vms/platformvm/locked/camino_lock.go (hunk 4 above)
Comment (evlekht) [reply vms/platformvm/locked/camino_lock.go:171]: Those and secpfx pkg ins/outs func are used in executor pkg inputsAreEqual and outsAreEqual funcs, which are used in FinishProposalsTx execution
I added a comments to Equal funcs

File: vms/platformvm/txs/executor/staker_tx_verification.go (hunk 5 above)
Comment (evlekht) [reply vms/platformvm/txs/executor/staker_tx_verification.go:302]: Thanks, there were one more file without header. Added to both

File: vms/platformvm/camino_service.go (hunk 7 above)
Comment (evlekht) [reply vms/platformvm/camino_service.go]: Renamed

File: vms/platformvm/dac/camino_change_base_fee_proposal.go (hunk 8 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_change_base_fee_proposal.go]: Commented all dac pkg struct fields

File: vms/platformvm/dac/camino_change_base_fee_proposal.go (hunk 9 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_change_base_fee_proposal.go]: Commented all dac pkg struct fields

File: vms/platformvm/dac/camino_change_base_fee_proposal.go (hunk 2 above)
Comment (evlekht) [reply vms/platformvm/dac/camino_change_base_fee_proposal.go:167]: Static analysis action complains about it, while my local linter isn't highlighting it (probably different versions)
Its here so ci will be green :)

**Comment by evlekht** [comment]
@c4t-ag
I added "version" without base fee proposal, see last 2 commits. First commit removes all base fee related stuff, then second adds it back. As you can see from first commit, it will unfortunately remove a lot of transaction tests, cause we need at least one proposal type in order to create them. And I don't want to add dummy proposal type, cause then we'll need to modify proposals visitors to include this type.

**Review by c4t-ag — APPROVED** [review]
Great work, go!

**Pull request merged**